File: code/packages/swift/dartmouth-basic-parser/Sources/DartmouthBasicParser/DartmouthBasicParser.swift
// ============================================================================
// DartmouthBasicParser.swift -- Dartmouth BASIC (1964) Grammar-Driven Parser
// ============================================================================
//
// This module parses Dartmouth BASIC source text into a generic ASTNode tree
// using the grammar-driven GrammarParser engine and the rules in
// dartmouth_basic.grammar.
//
// ============================================================================
// WHAT IS A GRAMMAR-DRIVEN PARSER?
// ============================================================================
//
// A traditional parser is hand-coded: the programmer writes one Swift function
// per grammar rule, and those functions call each other. This is called
// "recursive descent" parsing. It works, but every language needs its own
// carefully crafted code.
//
// A grammar-driven parser does something more powerful: it reads a text file
// that *describes* the grammar (dartmouth_basic.grammar), and interprets those
// rules at runtime. The same Swift engine that parses BASIC can also parse
// JSON, Python, or any language — just swap the .grammar file.
//
// The grammar file uses EBNF (Extended Backus-Naur Form) notation:
//
//   program = { line } ;
//   line = LINE_NUM [ statement ] NEWLINE ;
//   statement = let_stmt | print_stmt | ... ;
//   let_stmt = "LET" variable EQ expr ;
//   expr = term { ( PLUS | MINUS ) term } ;
//
// The GrammarParser reads these rules and builds a packrat memo table that
// efficiently matches the token stream. It uses Warth's algorithm to handle
// left-recursive rules without infinite loops.
//
// ============================================================================
// HOW DOES THE AST LOOK?
// ============================================================================
//
// For the program:
//   10 LET X = 5 + 3
//   20 END
//
// The AST looks like this (simplified):
//
//   ASTNode("program")
//   ├── ASTNode("line")
//   │   ├── Token(LINE_NUM, "10")
//   │   ├── ASTNode("statement")
//   │   │   └── ASTNode("let_stmt")
//   │   │       ├── Token(KEYWORD, "LET")
//   │   │       ├── ASTNode("variable")
//   │   │       │   └── Token(NAME, "x")
//   │   │       ├── Token(EQ, "=")
//   │   │       └── ASTNode("expr")
//   │   │           ├── ASTNode("term")
//   │   │           │   └── ASTNode("power")
//   │   │           │       └── ASTNode("unary")
//   │   │           │           └── ASTNode("primary")
//   │   │           │               └── Token(NUMBER, "5")
//   │   │           ├── Token(PLUS, "+")
//   │   │           └── ASTNode("term")
//   │   │               └── ... (wrapping NUMBER "3")
//   │   └── Token(NEWLINE, "\n")
//   └── ASTNode("line")
//       ├── Token(LINE_NUM, "20")
//       ├── ASTNode("statement")
//       │   └── ASTNode("end_stmt")
//       │       └── Token(KEYWORD, "END")
//       └── Token(NEWLINE, "\n")
//
// Each ASTNode carries a `ruleName` (the grammar rule that matched it),
// `children` (a mix of sub-nodes and leaf tokens), and source position info.
//
// ============================================================================
// EXPRESSION PRECEDENCE IN DARTMOUTH BASIC
// ============================================================================
//
// BASIC follows standard mathematical operator precedence, encoded as a
// cascade of grammar rules (lowest-to-highest binding):
//
//   1. expr  — addition (+) and subtraction (−), LEFT-associative
//              "1 + 2 + 3" → ((1 + 2) + 3)
//
//   2. term  — multiplication (*) and division (/), LEFT-associative
//              "2 * 3 * 4" → ((2 * 3) * 4)
//
//   3. power — exponentiation (^), RIGHT-associative
//              "2 ^ 3 ^ 2" → 2 ^ (3 ^ 2) = 512  (not (2^3)^2 = 64)
//              Right-associativity is modelled by the grammar rule:
//                power = unary [ CARET power ] ;
//              The recursive `power` on the right side chains right-to-left.
//
//   4. unary — unary minus (−)
//              "-X" → negate X
//
//   5. primary — atoms: numbers, function calls, variables, parenthesised exprs
//
// Truth table for operator priority (higher = binds tighter):
//
//   ^ (exponentiation)   : 5  (highest, right-associative)
//   - (unary minus)      : 4
//   * /                  : 3
//   + -                  : 2  (lowest arithmetic)
//   = < > <= >= <>       : 1  (comparison, only in IF)
//
// ============================================================================
// THE 17 STATEMENT TYPES
// ============================================================================
//
// Dartmouth BASIC 1964 has exactly 17 statement types:
//
//   1. LET     — variable assignment: LET X = expr
//   2. PRINT   — output: PRINT [list]
//   3. INPUT   — read from user: INPUT var, var, ...
//   4. IF      — conditional: IF expr relop expr THEN line_num
//   5. GOTO    — unconditional jump: GOTO line_num
//   6. GOSUB   — subroutine call: GOSUB line_num
//   7. RETURN  — return from subroutine: RETURN
//   8. FOR     — counted loop header: FOR var = from TO to [STEP step]
//   9. NEXT    — loop footer: NEXT var
//  10. END     — normal program end
//  11. STOP    — halt with message
//  12. REM     — remark/comment (content already suppressed by lexer)
//  13. READ    — read from DATA pool: READ var, var, ...
//  14. DATA    — literal data pool: DATA num, num, ...
//  15. RESTORE — reset DATA pointer: RESTORE
//  16. DIM     — array dimensioning: DIM name(size), ...
//  17. DEF     — define user function: DEF FNx(param) = expr
//
// ============================================================================
// THE LINE_NUM vs NUMBER DISAMBIGUATION
// ============================================================================
//
// The grammar rules `goto_stmt`, `gosub_stmt`, and `if_stmt` all reference
// `LINE_NUM` as the jump target token:
//
//   goto_stmt  = "GOTO"  LINE_NUM ;
//   gosub_stmt = "GOSUB" LINE_NUM ;
//   if_stmt    = "IF" expr relop expr "THEN" LINE_NUM ;
//
// However, the lexer's `relabelLineNumbers` pass only promotes the FIRST
// NUMBER on each source line to LINE_NUM. Jump targets like:
//
//   10 GOTO 50         — the "50" is mid-line, lexer emits NUMBER("50")
//   20 IF X > 0 THEN 100  — "100" is mid-line, lexer emits NUMBER("100")
//
// are emitted as NUMBER, not LINE_NUM.
//
// To bridge this gap, the parser registers a `relabelJumpTargets` pre-parse
// hook via `GrammarParser.addPreParse`. This hook walks the token list and
// promotes any NUMBER that immediately follows KEYWORD("GOTO"),
// KEYWORD("GOSUB"), or KEYWORD("THEN") to LINE_NUM.
//
// Why a pre-parse hook rather than a third lexer pass?
//   The lexer operates purely on the character stream and pattern matching.
//   It does not know the keyword context (whether a NUMBER follows GOTO or
//   appears in an arithmetic expression). The parser, after lexing, can see
//   the full token sequence and can apply context-sensitive relabeling.
//
//   A pre-parse hook is the clean integration point: it runs after lexing,
//   before the grammar rules are applied, and transforms the token list in
//   place. The GrammarParser.addPreParse API is designed for exactly this.
//
// Example:
//   Token stream after lexing: LINE_NUM("10") KEYWORD("GOTO") NUMBER("50") NEWLINE
//   After relabelJumpTargets:  LINE_NUM("10") KEYWORD("GOTO") LINE_NUM("50") NEWLINE
//
// ============================================================================
// USAGE
// ============================================================================
//
//     let ast = try DartmouthBasicParser.parse("10 LET X = 5\n20 END\n")
//     // ast.ruleName == "program"
//     // ast.children contains two "line" nodes
//
//     // Or from pre-lexed tokens:
//     let tokens = try DartmouthBasicLexer.tokenize(source)
//     let ast = try DartmouthBasicParser.parseTokens(tokens)
//
// ============================================================================

import Foundation
import GrammarTools
import Lexer
import Parser
import DartmouthBasicLexer

/// Dartmouth BASIC (1964) grammar-driven parser.
///
/// This struct is the public API for converting BASIC source text into an
/// ASTNode tree. Internally it uses:
///
///   1. `DartmouthBasicLexer.tokenize(_:)` — tokenize the source
///   2. `loadGrammar()` — load `dartmouth_basic.grammar` from the monorepo
///   3. `GrammarParser` — the grammar-driven parser engine, with a
///      `relabelJumpTargets` pre-parse hook to fix LINE_NUM vs NUMBER for
///      GOTO/GOSUB/IF-THEN targets
///
/// The resulting `ASTNode` has `ruleName = "program"` and a tree of children
/// that faithfully reflects every rule in dartmouth_basic.grammar.
///
/// All methods are static; the struct has no instance state, so it is
/// trivially `Sendable`.
///
public struct DartmouthBasicParser: Sendable {

    /// The version of this parser package.
    public static let version = "0.1.0"

    // =========================================================================
    // MARK: - Public API
    // =========================================================================

    /// Parse a Dartmouth BASIC source string into an AST.
    ///
    /// This is the main entry point. It lexes the source text, loads the
    /// grammar, and parses the resulting tokens.
    ///
    /// - Parameter source: The BASIC source code to parse (may be multi-line).
    /// - Returns: An `ASTNode` with `ruleName = "program"`.
    /// - Throws: Lexer errors (`LexerError`) or parser errors (`GrammarParseError`)
    ///           if the source is malformed.
    ///
    public static func parse(_ source: String) throws -> ASTNode {
        let tokens = try DartmouthBasicLexer.tokenize(source)
        return try parseTokens(tokens)
    }

    /// Parse a pre-lexed token stream into an AST.
    ///
    /// Use this when you already have a `[Token]` array — for example, if you
    /// want to inspect the token stream before parsing, or in tests that
    /// construct tokens manually.
    ///
    /// - Parameter tokens: A `[Token]` array from `DartmouthBasicLexer.tokenize(_:)`.
    /// - Returns: An `ASTNode` with `ruleName = "program"`.
    /// - Throws: `GrammarParseError` if the tokens do not form a valid program.
    ///
    public static func parseTokens(_ tokens: [Token]) throws -> ASTNode {
        let grammar = try loadGrammar()
        let parser = GrammarParser(tokens: tokens, grammar: grammar)

        // Register the pre-parse hook that promotes NUMBER tokens in jump-target
        // positions to LINE_NUM, so the grammar rules `goto_stmt = "GOTO" LINE_NUM`,
        // `gosub_stmt = "GOSUB" LINE_NUM`, and `if_stmt = "IF" ... "THEN" LINE_NUM`
        // match correctly.
        //
        // The lexer's relabelLineNumbers pass handles line-start positions.
        // This hook handles mid-line jump targets (GOTO 50, GOSUB 200, THEN 100).
        parser.addPreParse { tokens in
            tokens = relabelJumpTargets(tokens)
        }

        return try parser.parse()
    }

    /// Load and parse the Dartmouth BASIC parser grammar.
    ///
    /// The grammar file path is computed relative to this source file:
    ///
    ///   DartmouthBasicParser.swift         ← this file
    ///     Sources/DartmouthBasicParser/    (1) strip filename
    ///     dartmouth-basic-parser/          (2)
    ///     swift/                           (3)
    ///     packages/                        (4)
    ///     code/                            (5)  → grammars/dartmouth_basic.grammar
    ///
    /// The `#filePath` directive resolves to the compile-time path of this
    /// source file inside the monorepo. Walking up 6 levels reaches `code/`,
    /// then we descend into `grammars/`.
    ///
    /// - Returns: A `ParserGrammar` parsed from `dartmouth_basic.grammar`.
    /// - Throws: If the file cannot be read or the grammar cannot be parsed.
    ///
    public static func loadGrammar() throws -> ParserGrammar {
        let thisFile = #filePath
        var url = URL(fileURLWithPath: thisFile)
        // Walk up 6 path components:
        //   DartmouthBasicParser.swift  → Sources/DartmouthBasicParser
        //   Sources/DartmouthBasicParser → Sources
        //   Sources → dartmouth-basic-parser
        //   dartmouth-basic-parser → swift
        //   swift → packages
        //   packages → code
        for _ in 0..<6 {
            url = url.deletingLastPathComponent()
        }
        let grammarURL = url
            .appendingPathComponent("grammars")
            .appendingPathComponent("dartmouth_basic.grammar")

        let content = try String(contentsOf: grammarURL, encoding: .utf8)
        return try parseParserGrammar(source: content)
    }

    // =========================================================================
    // MARK: - Pre-parse Hook: relabelJumpTargets
    // =========================================================================

    /// Promote NUMBER tokens in jump-target position to LINE_NUM.
    ///
    /// The grammar rules for GOTO, GOSUB, and IF-THEN all reference LINE_NUM
    /// as the jump target:
    ///
    ///   goto_stmt  = "GOTO"  LINE_NUM ;
    ///   gosub_stmt = "GOSUB" LINE_NUM ;
    ///   if_stmt    = "IF" expr relop expr "THEN" LINE_NUM ;
    ///
    /// The lexer's `relabelLineNumbers` pass only promotes numbers at
    /// line-start. This function handles the mid-line jump targets:
    ///
    ///   Input:  ... KEYWORD("GOTO")  NUMBER("50") ...
    ///   Output: ... KEYWORD("GOTO")  LINE_NUM("50") ...
    ///
    ///   Input:  ... KEYWORD("GOSUB") NUMBER("200") ...
    ///   Output: ... KEYWORD("GOSUB") LINE_NUM("200") ...
    ///
    ///   Input:  ... KEYWORD("THEN")  NUMBER("100") ...
    ///   Output: ... KEYWORD("THEN")  LINE_NUM("100") ...
    ///
    /// Algorithm:
    ///   Walk tokens left-to-right, tracking whether the previous keyword
    ///   was GOTO, GOSUB, or THEN. When the current token is NUMBER and the
    ///   previous meaningful token was one of those keywords, relabel to LINE_NUM.
    ///
    /// - Parameter tokens: The token stream (modified in place).
    /// - Returns: A new token array with jump targets promoted.
    ///
    static func relabelJumpTargets(_ tokens: [Token]) -> [Token] {
        // Keywords whose immediately-following NUMBER is a line number, not
        // an arithmetic value. All three appear in the token stream with
        // uppercase values (thanks to the GrammarLexer's case-insensitive
        // keyword normalisation).
        let jumpPrecedingKeywords: Set<String> = ["GOTO", "GOSUB", "THEN"]

        var result: [Token] = []
        result.reserveCapacity(tokens.count)

        // `followsJumpKeyword` tracks whether the last KEYWORD token we saw
        // was GOTO, GOSUB, or THEN. If so, the next NUMBER token is a target.
        var followsJumpKeyword = false

        for token in tokens {
            switch token.type {
            case "KEYWORD":
                // Check whether this keyword introduces a jump target.
                followsJumpKeyword = jumpPrecedingKeywords.contains(token.value)
                result.append(token)

            case "NUMBER" where followsJumpKeyword:
                // This NUMBER immediately follows GOTO, GOSUB, or THEN.
                // Relabel it to LINE_NUM so the grammar rule matches.
                result.append(Token(
                    type: "LINE_NUM",
                    value: token.value,
                    line: token.line,
                    column: token.column,
                    flags: token.flags
                ))
                followsJumpKeyword = false

            default:
                // Any non-KEYWORD, non-NUMBER token resets the flag.
                // This handles edge cases like "IF X > 0 THEN 100":
                // Between THEN and 100 there are no tokens, so the flag
                // persists correctly. For more complex cases, we reset on
                // any intervening non-whitespace token.
                if token.type != "NEWLINE" && token.type != "EOF" {
                    // Only reset on real tokens (not structural separators).
                    // This is conservative — GOTO/GOSUB are always immediately
                    // followed by the target in valid BASIC programs.
                    followsJumpKeyword = false
                }
                result.append(token)
            }
        }

        return result
    }
}

File: code/packages/elixir/dartmouth_basic_parser/lib/coding_adventures/dartmouth_basic_parser.ex
defmodule CodingAdventures.DartmouthBasicParser do
  @moduledoc """
  Dartmouth BASIC Parser — Thin wrapper around the grammar-driven parser engine.

  ## A Brief History: Parsing BASIC

  When John Kemeny and Thomas Kurtz designed Dartmouth BASIC in 1964, parsing
  was a manual, bespoke affair. Each language had its own hand-written parser
  because the field of formal language theory was young. Chomsky had published
  his hierarchy only in 1956; Knuth's LR parsing paper was still two years
  away (1965). The BASIC system on the GE-225 used a simple top-down recursive
  scan because BASIC's grammar is deliberately unambiguous: every statement
  begins with a distinct keyword, and every expression uses the classic
  precedence cascade of arithmetic.

  Today we can parse BASIC using a *grammar-driven* engine: a single file,
  `dartmouth_basic.grammar`, describes the complete syntax in EBNF notation,
  and a reusable parser engine interprets it at runtime. This package is a
  thin adapter that wires together:

    1. The **Dartmouth BASIC lexer** (`dartmouth_basic_lexer`) — converts raw
       source text into a typed token stream with LINE_NUM, KEYWORD, NAME, etc.

    2. The **grammar-driven parser engine** (`parser`) — reads
       `dartmouth_basic.grammar` and applies it to the token stream using
       recursive-descent with packrat memoization.

  The result is an **Abstract Syntax Tree (AST)** — a tree of `ASTNode`
  structs whose `rule_name` fields mirror the grammar rules (`program`, `line`,
  `statement`, `let_stmt`, `expr`, etc.).

  ## The Grammar at a Glance

  Dartmouth BASIC 1964 has 17 statement types:

  | Statement   | Example                      | Purpose                  |
  |-------------|------------------------------|--------------------------|
  | LET         | `10 LET X = 5`               | Variable assignment      |
  | PRINT       | `10 PRINT X, Y`              | Output to terminal       |
  | INPUT       | `10 INPUT A, B`              | Read values from user    |
  | IF...THEN   | `10 IF X > 0 THEN 100`       | Conditional branch       |
  | GOTO        | `10 GOTO 50`                 | Unconditional jump       |
  | GOSUB       | `10 GOSUB 200`               | Subroutine call          |
  | RETURN      | `200 RETURN`                 | Return from subroutine   |
  | FOR         | `10 FOR I = 1 TO 10`         | Loop start               |
  | NEXT        | `30 NEXT I`                  | Loop end                 |
  | END         | `99 END`                     | Normal termination       |
  | STOP        | `99 STOP`                    | Halt with message        |
  | REM         | `10 REM A COMMENT`           | Remark / comment         |
  | READ        | `10 READ X, Y`               | Read from DATA pool      |
  | DATA        | `20 DATA 1, 2, 3`            | Declare data pool values |
  | RESTORE     | `30 RESTORE`                 | Reset DATA pointer       |
  | DIM         | `10 DIM A(100)`              | Dimension an array       |
  | DEF         | `10 DEF FNA(X) = X * X`      | Define a function        |

  Expressions support:
  - `+`, `-` (left-associative, lowest precedence)
  - `*`, `/` (left-associative, medium precedence)
  - `^` (right-associative exponentiation: `2^3^2` = 512, not 64)
  - unary minus (`-X`)
  - parenthesised subexpressions (`(X + 1)`)
  - built-in functions: SIN, COS, TAN, ATN, EXP, LOG, ABS, SQR, INT, RND, SGN
  - user-defined functions: FNA through FNZ
  - scalar variables (A–Z, A0–Z9)
  - array elements (A(I))

  ## Pipeline Position

  ```
  BASIC source text  (e.g., "10 LET X = 42\\n20 PRINT X\\n")
        │
        ▼  DartmouthBasicLexer.tokenize/1
  ┌──────────────────────────────────┐
  │   dartmouth_basic_lexer          │
  │   dartmouth_basic.tokens         │
  └──────────────────────────────────┘
        │  [{type, value, line, column}, ...]
        ▼  DartmouthBasicParser.parse/1
  ┌──────────────────────────────────┐
  │   dartmouth_basic_parser         │  ← this package
  │   dartmouth_basic.grammar        │
  └──────────────────────────────────┘
        │  %ASTNode{rule_name: "program", ...}
        ▼
  ┌──────────────────────────────────┐
  │   dartmouth_basic_compiler       │
  └──────────────────────────────────┘
  ```

  ## Usage

      {:ok, ast} = CodingAdventures.DartmouthBasicParser.parse_source("10 PRINT \\"HELLO, WORLD\\"\\n20 END\\n")
      ast.rule_name  # => "program"

      # Or tokenize first, then parse:
      {:ok, tokens} = CodingAdventures.DartmouthBasicLexer.tokenize(source)
      {:ok, ast}    = CodingAdventures.DartmouthBasicParser.parse(tokens)

  ## Grammar Caching

  The `dartmouth_basic.grammar` file is read once and cached in `:persistent_term`
  (a BEAM-level key-value store for "write once, read many" immutable data).
  Unlike ETS (which copies terms on every access), `persistent_term` returns a
  direct reference with zero copy overhead. The first call pays the file I/O cost;
  every subsequent call is effectively free.
  """

  alias CodingAdventures.GrammarTools.ParserGrammar
  alias CodingAdventures.DartmouthBasicLexer
  alias CodingAdventures.Parser.{GrammarParser, ASTNode}

  # ---------------------------------------------------------------------------
  # Grammar path resolution
  # ---------------------------------------------------------------------------
  #
  # __DIR__ is the compile-time directory of THIS source file:
  #   .../code/packages/elixir/dartmouth_basic_parser/lib/coding_adventures
  #
  # Walking up with ".." five times reaches code/:
  #   (1) lib/coding_adventures  →  lib
  #   (2) lib                    →  dartmouth_basic_parser
  #   (3) dartmouth_basic_parser →  elixir
  #   (4) elixir                 →  packages
  #   (5) packages               →  code
  #   then append "grammars"     →  code/grammars
  #
  # Path.expand/1 resolves all ".." components to an absolute path so that
  # File.read! works regardless of the current working directory at runtime.
  @grammars_dir Path.join([__DIR__, "..", "..", "..", "..", "..", "grammars"])
                |> Path.expand()

  # ---------------------------------------------------------------------------
  # Public API
  # ---------------------------------------------------------------------------

  @doc """
  Parse a pre-tokenized list of Dartmouth BASIC tokens into an AST.

  Accepts a list of `%Token{}` structs (as produced by `DartmouthBasicLexer.tokenize/1`)
  and returns an AST rooted at the `"program"` rule.

  ## When to use `parse/1` vs `parse_source/1`

  - Use `parse_source/1` for the common case: raw BASIC source → AST.
  - Use `parse/1` when you have already tokenised the source (e.g., in a
    pipeline where you need to inspect or transform the tokens before parsing).

  ## Return value

  - `{:ok, %ASTNode{rule_name: "program", ...}}` on success.
  - `{:error, message}` if parsing fails (invalid syntax).

  ## Example

      {:ok, tokens} = DartmouthBasicLexer.tokenize("10 LET X = 5\\n")
      {:ok, ast}    = DartmouthBasicParser.parse(tokens)
      ast.rule_name  # => "program"

  """
  @spec parse([CodingAdventures.Lexer.Token.t()]) :: {:ok, ASTNode.t()} | {:error, String.t()}
  def parse(tokens) do
    grammar = get_grammar()
    GrammarParser.parse(tokens, grammar)
  end

  @doc """
  Tokenize Dartmouth BASIC source code and parse it into an AST in one call.

  This is the primary entry point for callers who have raw BASIC source text.
  Internally it calls `DartmouthBasicLexer.tokenize/1` followed by
  `GrammarParser.parse/2`.

  ## Return value

  - `{:ok, %ASTNode{rule_name: "program", ...}}` on success.
  - `{:error, message}` if lexing or parsing fails.

  ## Example

      {:ok, ast} = DartmouthBasicParser.parse_source("10 PRINT \\"HELLO, WORLD\\"\\n20 END\\n")
      ast.rule_name  # => "program"

  ## Multi-line programs

      # A multi-line program spanning several lines:
      src = "10 FOR I = 1 TO 5\\n20 PRINT I\\n30 NEXT I\\n40 END\\n"
      {:ok, ast} = DartmouthBasicParser.parse_source(src)

  """
  @spec parse_source(String.t()) :: {:ok, ASTNode.t()} | {:error, String.t()}
  def parse_source(source) do
    case DartmouthBasicLexer.tokenize(source) do
      {:ok, tokens} -> parse(tokens)
      {:error, msg} -> {:error, msg}
    end
  end

  @doc """
  Parse the `dartmouth_basic.grammar` file and return the `ParserGrammar`.

  Returns a `%ParserGrammar{}` struct containing all grammar rules. You can
  inspect it to enumerate the rules, check alternation order, or verify that
  the grammar loaded correctly. The grammar drives the parser engine in
  `GrammarParser.parse/2`.

  This function always reads fresh from disk — it does NOT return the cached
  copy. Use it for introspection and testing. For production parsing, prefer
  `parse/1` or `parse_source/1`, which use the cached grammar.

  ## Example

      grammar = DartmouthBasicParser.create_parser()
      rule_names = Enum.map(grammar.rules, & &1.name)
      # => ["program", "line", "statement", "let_stmt", ...]

  """
  @spec create_parser() :: ParserGrammar.t()
  def create_parser do
    grammar_path = Path.join(@grammars_dir, "dartmouth_basic.grammar")
    {:ok, grammar} = ParserGrammar.parse(File.read!(grammar_path))
    grammar
  end

  # ---------------------------------------------------------------------------
  # Grammar caching (private)
  # ---------------------------------------------------------------------------
  #
  # The grammar is cached in :persistent_term — a BEAM key-value store designed
  # for data that is written once and read many times. Unlike ETS, persistent_term
  # returns a direct reference to the stored term (no copying). This means the
  # grammar struct is shared across all processes with zero overhead.
  #
  # Key: {CodingAdventures.DartmouthBasicParser, :grammar}
  #   — Using the module name as part of the key avoids collisions with other
  #     packages that also use persistent_term (e.g., DartmouthBasicLexer).
  defp get_grammar do
    case :persistent_term.get({__MODULE__, :grammar}, nil) do
      nil ->
        # First call: load from disk and cache.
        grammar = create_parser()
        :persistent_term.put({__MODULE__, :grammar}, grammar)
        grammar

      grammar ->
        # Subsequent calls: return cached grammar directly (zero copy).
        grammar
    end
  end
end
